### ISSUE TITLE
Allow luaref for `indentexpr`, `foldexpr` and other exprs

### ISSUE THREAD

**Issue by stasjok** [neovim/neovim#18132]
It is very cool that we can now use a lua functions in mappings, commands, autocmds directly without vimscript bridge. Would it be possible to implement an ability to use lua functions directly in `indentexpr`? I know that there is a `v:lua`, but it is still converted via vimscript, right?

**Comment by clason** [comment]
Should also include the various `*func` options (like `omnifunc`, `tagfunc`, etc.), which share the basic plumbing.

**Comment by justinmk** [comment]
Duplicate of https://github.com/neovim/neovim/issues/14157

**Comment by justinmk** [comment]
See https://github.com/neovim/neovim/issues/11306 for proposed interface.

Copying https://github.com/neovim/neovim/issues/14157#issuecomment-1320787927 here for reference:

> There is a way to set `*func` options to Lua functions without adding global functions:
> 
> `'operatorfunc'`:
> 
> ```lua
> local set_opfunc = vim.fn[vim.api.nvim_exec([[
>   func s:set_opfunc(val)
>     let &opfunc = a:val
>   endfunc
>   echon get(function('s:set_opfunc'), 'name')
> ]], true)]
> ```
> 
> `'tagfunc'` for current buffer:
> 
> ```lua
> local set_tagfunc = vim.fn[vim.api.nvim_exec([[
>   func s:set_tagfunc(val)
>     let &l:tagfunc = a:val
>   endfunc
>   echon get(function('s:set_tagfunc'), 'name')
> ]], true)]
> ```
> 
> `'tagfunc'` for non-current buffer:
> 
> ```lua
> local set_tagfunc = vim.fn[vim.api.nvim_exec([[
>   func s:set_tagfunc(buf, val)
>     call setbufvar(a:buf, '&tagfunc', string(a:val))
>   endfunc
>   echon get(function('s:set_tagfunc'), 'name')
> ]], true)]
> ```

**Comment by justinmk** [comment]
See https://github.com/neovim/neovim/issues/11306 for proposed interface.

**Comment by lewis6991** [comment]
The suggestions in #11306 imply the vimscript bridge is still required? Ideally options (specifically) can have a special syntax for indicating Lua expressions so we can avoid typical conversion completely.

**Comment by justinmk** [comment]
> The suggestions in #11306 imply the vimscript bridge is still required? Ideally options (specifically) can have a special syntax for indicating Lua expressions so we can avoid typval conversion completely.

Couldn't `set =# <Lua>` skip typval conversion? Or what do you suggest?

**Comment by lewis6991** [comment]
Yes it could. Just trying to clarify and align expectations. Especially since the OP mentions it next to `let`.

**Comment by stasjok** [comment]
To be honest I expected something like `vim.bo.indentexpr = func` and `vim.api.nvim_set_option_value("indentexpr", func, {buf= 1, local = true})` (or `vim.api.nvim_set_option_value("indentexpr", "", {buf= 1, local = true, callback = func})`) since it's not allowed to change types for api function arguments) where `func` is a closure (can use it's up-values).

**Comment by justinmk** [comment]
> To be honest I expected something like `vim.bo.indentexpr = func`

Yes that's also part of the story. But I think the changes to internals are the main cost, which applies to both cases.
